### DISCUSSION TITLE
When will component licensing be available again?

### DISCUSSION THREAD

**Discussion by cadilhac** [dotnet/winforms#5080] [Q&A] [answered]
My company sells 2 windows forms components for .Net. They rely on the licensing engine (with licenses.licx file). 
My customers now expect a version of my products for .Net Core3 / 5 / 6. But I'm stuck because licensing is not possible anymore. When will this be implemented again (in the old way or in a new way)? We see new versions of .Net since Windows forms was announced for .Net core, but there seems to be no progress on this licensing side... This really hurts my company.
Thanks.

**Comment by AraHaan** [comment]
I mean what's the point? From Day 1 of .NET Framework to .NET 5 / 6 they could have just as easily used ILSpy to get the code.

An even better option is to just add some sort of ``product key`` (that is read from a file so developer using the components and end users using the programs that use it) system in order to actually use the components and have them register for it and read the licensing stuff online where they get the key that then tells them what they *can* / *cant* do with the components.

As for the System, I would make it where they expire so then they got to re-read the information in order to obtain a new "key" to enter to continue use every so often as people do forget things they read about it as well.

**Reply by weltkante** [reply]
When implemented properly the Desktop Framework licensing mechanism is safe and not circumventable by looking at the code via ILSpy. If the license status is sourced by a cryptographic source (e.g. asymetric encryption via RSA for reading out encrypted license information and matching them to the machine its running on) no looking at the code will provide you the cryptographic keys to reproduce the values a licensed machine has.

That said many component developers get it wrong and implemented it incorrectly, so yes, you could circumvent those bad implementations. That doesn't mean there is no need for having a proper licensing infrastructure.

**Reply by AraHaan** [reply]
That is entirely true.

**Comment by RussKie** [comment, marked as answer]
https://github.com/dotnet/winforms/issues/3316#issuecomment-964629843

From @merriemcgaw 
> _There are no plans to implement the licx file mechanism for licensing. The current thinking internally is that most software vendors are creating their own licensing mechanism, and that the licx file method was not adequate for most modern scenarios. The recommendation is to either invest in a cloud-based licensing approach or create a system specific to your component and your component needs. Microsoft does not have plans to invest in a licensing model provided by Microsoft in .NET. If you have any questions, you can reach out to me or @OliaG directly, and we'd be happy to pass along feedback up our management chain._
